src/rust/cryptography-x509-validation/src/lib.rs

```diff
@@ -11,6 +11,349 @@ pub mod policy;
+                        pattern, name,
+                    ))),
+                    (None, _) => Err(ValidationError::Other(format!(
+                        "malformed IP name constraints: {:?}",
+                        pattern
+                    ))),
+                }
+            }
+            _ => Ok(Skipped),
+        }
+    }
+
+    /// Apply the current name constraints (including those in the specified
+    /// extensions, if any) to the accumulated SAN set.
+    ///
+    /// On success (no constraint violations found), the new constraints
+    /// are additionally added to the name constraint set for future checks.
+    fn apply_and_accumulate(&mut self, extensions: &Extensions<'a>) -> Result<(), ValidationError> {
+        let new_constraints = match extensions.get_extension(&NAME_CONSTRAINTS_OID) {
+            Some(nc) => Some(nc.value::<NameConstraints<'a>>()?),
+            None => None,
+        };
+
+        for san in &self.sans {
+            // If there are no applicable constraints, the SAN is considered valid so the default is true.
+            let mut permit = true;
+            for nc in self.name_constraints.iter().chain(new_constraints.iter()) {
+                if let Some(permitted_subtrees) = &nc.permitted_subtrees {
+                    for p in permitted_subtrees.unwrap_read().clone() {
+                        let status = self.apply_inner(&p.base, san)?;
+                        if status.is_applied() {
+                            permit = status.is_match();
+                            if permit {
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                if !permit {
+                    return Err(ValidationError::Other(
+                        "no permitted name constraints matched SAN".into(),
+                    ));
+                }
+            }
+
+            for nc in self.name_constraints.iter().chain(new_constraints.iter()) {
+                if let Some(excluded_subtrees) = &nc.excluded_subtrees {
+                    for e in excluded_subtrees.unwrap_read().clone() {
+                        let status = self.apply_inner(&e.base, san)?;
+                        if status.is_match() {
+                            return Err(ValidationError::Other(
+                                "excluded name constraint matched SAN".into(),
+                            ));
+                        }
+                    }
+                }
+            }
+        }
+
+        if let Some(new_constraints) = new_constraints {
+            self.name_constraints.push(new_constraints);
+        }
+
+        Ok(())
+    }
+}
+
+pub struct Intermediates<'a>(HashSet<Certificate<'a>>);
+
+impl<'a> Intermediates<'a> {
+    fn new(intermediates: impl IntoIterator<Item = Certificate<'a>>) -> Self {
+        Self(intermediates.into_iter().collect())
+    }
+}
```

Comment:
Flagging for review: now that I've removed the self-signed special casing, this newtype has no custom invariants. It can be removed if you don't see the need for an additional type here 🙂

Reply:
Yeah, I favor dropping it.